docs/backend-system/building-plugins-and-modules/08-migrating.md

```diff
@@ -6,4 +6,223 @@ sidebar_label: Migration Guide
+
+The kubernetes backend plugin only supports one `ClusterSupplier` at this time but let's look at how we could add support for multiple suppliers using extension points. This allows users to install several modules that add their own `ClusterSupplier` implementations to the plugin like this:
+
+```ts
+backend.add(kubernetesPlugin());
+backend.add(GoogleContainerEngineModule());
+backend.add(EKSModule());
+```
+
+Now let's look at how to implement this with extension points. First we need to define the extension point itself. As the ExtensionPoint will be used by other modules we need to export its common practice to export these from a shared package so that they can be imported by other modules and plugins.
+
+We'll go ahead and create a `@backstage/plugin-kubernetes-node` package for this and from there we'll export the extension point.
+
+```ts
+import { createExtensionPoint } from '@backstage/backend-plugin-api';
+
+export interface KubernetesClusterSupplierExtensionPoint {
+  addClusterSupplier(supplier: KubernetesClustersSupplier): void;
+}
+
+/**
+ * An extension point that allows other plugins to add catalog processors.
+ * @public
+ */
+export const kubernetesClustersSupplierExtensionPoint =
+  createExtensionPoint<KubernetesClusterSupplierExtensionPoint>({
+    id: 'kubernetes.cluster-supplier',
+  });
+```
+
+Now we can use this extension point in the kubernetes backend plugin to register the ExtensionPoint for modules to use.
+
+```ts
+import { kubernetesClustersSupplierExtensionPoint } from '@backstage/plugin-kubernetes-node';
+
+// Our internal implementation of the extensionPoint which we do not need to export.
+class ClusterSupplier implements KubernetesClustersSupplier {
+  private clusterSuppliers: KubernetesClustersSupplier[] = [];
+
+  // This method is private and only used internally to retrieve the registered suppliers.
+  getClusterSuppliers(): KubernetesClustersSupplier[] {
+    return this.clusterSuppliers;
+  }
+
+  addClusterSupplier(supplier: KubernetesClustersSupplier) {
+    this.clusterSuppliers.push(supplier);
+  }
+}
+
+export const kubernetesPlugin = createBackendPlugin({
+  id: 'kubernetes',
+  register(env) {
+    const extensionPoint = new ClusterSupplier();
+    // We register the extensionPoint with the backend which allows modules to register their own ClusterSupplier.
+    env.registerExtensionPoint(
+      kubernetesClustersSupplierExtensionPoint,
+      extensionPoint,
+    );
+
+    env.registerInit({
+      deps: {
+        ... omitted ...
+      },
+      async init({ config, logger, catalogApi, discovery, http }) {
+        const { router } = await KubernetesBuilder.createBuilder({
+          config,
+          logger,
+          catalogApi,
+          discovery,
+        })
+          // We pass in the all the registered suppliers to the builder.
+          .setClusterSuppliers(...extensionPoint.getClusterSuppliers())
```

Comment:
This is `setClusterSupplier` in the existing implementation, you can only have one. I'm thinking it would be fine to highlight that pattern, and basically throw an error if multiple modules try to set their own cluster suppliers?

Reply:
yeah I can do that instead 👍